service/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java

```diff
@@ -1,93 +1,119 @@
+    private final TagDao tagDao;
+    private final ModelMapper modelMapper;
+
+    public CertificateServiceImpl(CertificateDao certificateDAO, TagDao tagDao, ModelMapper modelMapper) {
         this.certificateDAO = certificateDAO;
-        this.certificateDTOMapper = certificateDTOMapper;
-        this.tagService = tagService;
-        this.queryProvider = queryProvider;
-        this.queryParametersProvider = queryParametersProvider;
+        this.tagDao = tagDao;
+        this.modelMapper = modelMapper;
     }
 
     @Override
     @Transactional
-    public Integer update(CertificateDTO certificate) {
-        return certificateDAO.update(certificateDTOMapper.toEntity(certificate));
+    public CertificateDTO update(CertificateDTO dto) {
+        Certificate existing = certificateDAO.readById(dto.getId())
+                .orElseThrow(() -> new IdNotExistException(dto.getId().toString()));
+        updateFieldsFromDto(existing, dto);
+        existing = certificateDAO.update(existing).orElseThrow(UpdateEntityInternalException::new);
+        return modelMapper.map(existing, CertificateDTO.class);
     }
 
     @Override
-    public List<CertificateDTO> searchByParameters(String tagName, String certificateName, String description, String sortBy, String order) {
-        String queryPart = queryProvider.getCertificateSearchQuery(tagName, certificateName, description, sortBy, order);
-        List<String> queryParameters = queryParametersProvider.getCertificateParameters(tagName, certificateName, description);
-
-        return certificateDAO.searchByParameters(queryPart, queryParameters)
-                .stream()
-                .map(certificate -> certificateDTOMapper.toDto(certificate, tagService.readByCertificateId(certificate.getId())))
+    public List<CertificateDTO> searchByParameters(SearchParameter searchParameter, int page, int size) {
+        List<Certificate> certificates = certificateDAO.searchByParameters(searchParameter, page, size);
+        return certificates.stream()
+                .map(certificate -> modelMapper.map(certificate, CertificateDTO.class))
                 .collect(Collectors.toList());
     }
 
     @Override
     @Transactional
-    public Long create(CertificateDTO dto) {
-        // insert certificate
-        Certificate certificateEntity = certificateDTOMapper.toEntity(dto);
-        Long certificateId = certificateDAO.create(certificateEntity);
-
-        // insert tags
-        List<TagDTO> tagsDTO = dto.getTags();
-        tagsDTO.forEach(tagDTO -> {
-            Long tagId = tagService.readByName(tagDTO.getName()).map(Tag::getId).orElseGet(() -> tagService.create(tagDTO));
-            certificateDAO.createCertificateHasTag(certificateId, tagId);
-        });
-        return certificateId;
+    public CertificateDTO create(CertificateDTO dto) {
+        Certificate certificate = modelMapper.map(dto, Certificate.class);

```

Comment:
From my side, I would prefer certifacateDto name instead simple dto for higher readability

Reply:
thanks, renamed